In [ ]:
#!pip install beautifulsoup4

In [35]:
import requests
from bs4 import BeautifulSoup
import re

In [5]:
result = requests.get("https://www.google.com/")

In [9]:
result.status_code

200

200 is an OK response and means the response was valid

In [8]:
result.content

b'<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="en-ZA"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="L+89trQACCI0i3d1kchdhA==">(function(){window.google={kEI:\'zQXnXL2iDaXgkgW4xbr4Bw\',kEXPI:\'0,1353747,57,51,1907,2422,697,528,590,140,224,1575,1257,1894,58,321,206,1017,735,157,295,162,2332298,329477,1294,12383,4855,32692,15247,867,12163,6381,3335,2,2,4609,2192,363,3320,5505,224,2212,266,5107,575,835,284,2,578,728,2432,58,2,4,1297,4323,4967,774,2251,1406,3337,1146,5,2,2,1494,255,2809,3601,669,1050,1808,1397,81,7,3,488,620,29,1395,3610,5299,1288,2,4007,796,101,1119,38,920,753,120,1217,1364,1611,1582,1154,3061,2,631,2562,2,4,2,670,45,4147,112,398,125,1161,1448,10,620,1139,1089,656,338,1592,389,228,2,961,198,777,1,2,367,1016,300,705,756,98,36,2,354,29,400,992,1107,10,168,9,108,187,831,139,96,810,450,174,9

So clearly this prints the entire body of the HTML for the page

We can also request things like the headers of the page

In [10]:
result.headers

{'Date': 'Thu, 23 May 2019 20:42:53 GMT', 'Expires': '-1', 'Cache-Control': 'private, max-age=0', 'Content-Type': 'text/html; charset=ISO-8859-1', 'P3P': 'CP="This is not a P3P policy! See g.co/p3phelp for more info."', 'Content-Encoding': 'gzip', 'Server': 'gws', 'X-XSS-Protection': '0', 'X-Frame-Options': 'SAMEORIGIN', 'Set-Cookie': '1P_JAR=2019-05-23-20; expires=Sat, 22-Jun-2019 20:42:53 GMT; path=/; domain=.google.com, NID=184=JuaaGIOQLNEWEnJVR8Ly9q3messfELkJDnQpQqkdyouE7imvg_oAVA7nBoBMP6KQbIalMDIJ3-S4gSh7R7B6fXcRFXH2wVKVYVO7R14T79YzLK3OJ1dFl9D8JvjqlocK4P5lbz5NdOY-b2dlp8BA3cvtx4dTZ1bUjLcg8pmkANE; expires=Fri, 22-Nov-2019 20:42:53 GMT; path=/; domain=.google.com; HttpOnly', 'Alt-Svc': 'quic=":443"; ma=2592000; v="46,44,43,39"', 'Transfer-Encoding': 'chunked'}

Now we can start extracting some of the information on this page

In [15]:
soup = BeautifulSoup(result.content, 'html.parser')

In [16]:
links = soup.find_all("a")

In [17]:
links

[<a class="gb1" href="https://www.google.co.za/imghp?hl=en&amp;tab=wi">Images</a>,
 <a class="gb1" href="https://maps.google.co.za/maps?hl=en&amp;tab=wl">Maps</a>,
 <a class="gb1" href="https://play.google.com/?hl=en&amp;tab=w8">Play</a>,
 <a class="gb1" href="https://www.youtube.com/?gl=ZA&amp;tab=w1">YouTube</a>,
 <a class="gb1" href="https://news.google.co.za/nwshp?hl=en&amp;tab=wn">News</a>,
 <a class="gb1" href="https://mail.google.com/mail/?tab=wm">Gmail</a>,
 <a class="gb1" href="https://drive.google.com/?tab=wo">Drive</a>,
 <a class="gb1" href="https://www.google.co.za/intl/en/about/products?tab=wh" style="text-decoration:none"><u>More</u> »</a>,
 <a class="gb4" href="http://www.google.co.za/history/optout?hl=en">Web History</a>,
 <a class="gb4" href="/preferences?hl=en">Settings</a>,
 <a class="gb4" href="https://accounts.google.com/ServiceLogin?hl=en&amp;passive=true&amp;continue=https://www.google.com/" id="gb_70" target="_top">Sign in</a>,
 <a href="/advanced_search?hl=en-Z

In [23]:
links_with_about = [link for link in soup.find_all("a") if "About" in link.text]

In [25]:
links_with_about

[<a href="/intl/en/about.html">About Google</a>]

Okay for this tutorial we want to try and use these libraries to get some data on the interest rates within different developing countries

To do that we will pick a website with relevant information...

After a quick google I came across the following;

https://tradingeconomics.com/country-list/interest-rate

In [26]:
int_rates = requests.get("https://tradingeconomics.com/country-list/interest-rate")

In [27]:
soup = BeautifulSoup(int_rates.content, 'html.parser')

In [28]:
soup


<!DOCTYPE doctype html>

<html>
<head id="ctl00_Head1"><meta charset="utf-8"/><title>
	Interest Rate - Countries - List
</title><meta content="This page displays a table with actual values, consensus figures, forecasts, statistics and historical data charts for - Interest Rate. This page provides values for Interest Rate reported in several countries. The table has current values for Interest Rate, previous releases, historical highs and record lows, release frequency, reported unit and currency plus links to historical data charts." id="metaDesc" name="description"/><meta content="Country List,Interest Rate,data,chart,actual,historical,values,calendar,forecast,graph,2015,2016,2017" id="metaKeyword" name="keywords"/><meta content="width=device-width,minimum-scale=1,initial-scale=1" name="viewport"/><meta content="#333333" name="theme-color"/><meta content="780337348699873" property="fb:app_id"/><meta content="100008712870955" property="fb:admins"/><link href="https://d3fy651gv2fhd3.cl

Okay fuck, this is a crazy mess...

How do I get at this html table here?

Try more turorials;
https://www.google.co.za/search?q=scraping+data+from+html+tables+python&oq=scraping+data+from+tables&aqs=chrome.3.69i57j0l4.8712j0j7&client=ubuntu&sourceid=chrome&ie=UTF-8#kpvalbx=1


In [29]:
soup.find_all('tr')

[<tr>
 <th style="cursor: pointer">Country</th>
 <th data-sortable="true" style="text-align: right; cursor: pointer">Last </th>
 <th></th>
 <th data-sortable="true" style="text-align: right; cursor: pointer">Previous </th>
 <th class="hidden-sm hidden-xs" data-sortable="true" style="cursor: pointer; text-align:center">Range</th>
 <th class="hidden-xs" data-sortable="false"></th>
 </tr>, <tr class="datatable-row">
 <td style="max-width: 120px; overflow: hidden; padding-left: 10px; text-align: left; font-weight: 600;">
 <a href="/argentina/interest-rate">
                                         Argentina
                                     </a></td>
 <td data-value="71.22" style="text-align: right;">
                                     71.22
                                 </td>
 <td style="text-align: right;"><span>May/19</span></td>
 <td class="hidden-sm hidden-xs" style="text-align: right;">71.44</td>
 <td class="hidden-sm hidden-xs" style="text-align: center;">1390 : 1.2</td>
 <t

This looks juicy!

I can see the words "argintina interest rate and some data-value= 71.22

Let's look this up manually to confirm that we are on to something...

Dope that works!

We just have to extract all the relevant details we want exactly, making sure we don't confuse the different columns.

Or better pull columns and store all the data into a csv file...

So how would we do this? We can try the following psuedo code;

1. pull all tr responses that end with 'datatable-row'
- pull each td value witin that row as elements of a row
- clean each element and store them comma delim

In [30]:
table_contents = soup.find_all('tr')

In [33]:
row_contents = soup.find_all('td')

In [34]:
row_contents

[<td style="max-width: 120px; overflow: hidden; padding-left: 10px; text-align: left; font-weight: 600;">
 <a href="/argentina/interest-rate">
                                         Argentina
                                     </a></td>,
 <td data-value="71.22" style="text-align: right;">
                                     71.22
                                 </td>,
 <td style="text-align: right;"><span>May/19</span></td>,
 <td class="hidden-sm hidden-xs" style="text-align: right;">71.44</td>,
 <td class="hidden-sm hidden-xs" style="text-align: center;">1390 : 1.2</td>,
 <td class="hidden-xs">
                                     %
                                 </td>,
 <td style="max-width: 120px; overflow: hidden; padding-left: 10px; text-align: left; font-weight: 600;">
 <a href="/yemen/interest-rate">
                                         Yemen
                                     </a></td>,
 <td data-value="27.00" style="text-align: right;">
                          

In [116]:
country = [re.findall(r"[A-z]+",row_contents[i].text) for i in range(len(row_contents)) if i%6 == 0]

In [117]:
country

[['Argentina'],
 ['Yemen'],
 ['Suriname'],
 ['Turkey'],
 ['Venezuela'],
 ['Haiti'],
 ['Iran'],
 ['Ukraine'],
 ['Sierra', 'Leone'],
 ['Sudan'],
 ['Ghana'],
 ['Uzbekistan'],
 ['Angola'],
 ['Egypt'],
 ['Tajikistan'],
 ['Mozambique'],
 ['Malawi'],
 ['Nigeria'],
 ['Seychelles'],
 ['Gambia'],
 ['Guinea'],
 ['Liberia'],
 ['Pakistan'],
 ['Djibouti'],
 ['Mongolia'],
 ['Zambia'],
 ['Belarus'],
 ['Lebanon'],
 ['Myanmar'],
 ['Uganda'],
 ['Madagascar'],
 ['Zimbabwe'],
 ['Congo'],
 ['Kazakhstan'],
 ['Kenya'],
 ['Sao', 'Tome', 'and', 'Principe'],
 ['Azerbaijan'],
 ['Ecuador'],
 ['Mexico'],
 ['Sri', 'Lanka'],
 ['Russia'],
 ['Tunisia'],
 ['Bhutan'],
 ['Ethiopia'],
 ['Maldives'],
 ['Tanzania'],
 ['Namibia'],
 ['South', 'Africa'],
 ['Swaziland'],
 ['Lesotho'],
 ['Brazil'],
 ['Georgia'],
 ['Mauritania'],
 ['Moldova'],
 ['Nepal'],
 ['Papua', 'New', 'Guinea'],
 ['Vietnam'],
 ['Bangladesh'],
 ['India'],
 ['Indonesia'],
 ['Armenia'],
 ['Honduras'],
 ['Brunei'],
 ['Cape', 'Verde'],
 ['Dominican', 'Republic'],


In [118]:
last = [re.findall(r"[0-9]+\.[0-9]+",row_contents[i].text) for i in range(len(row_contents)) if i%6 == 1]

In [119]:
last

[['71.22'],
 ['27.00'],
 ['25.00'],
 ['24.00'],
 ['22.40'],
 ['20.00'],
 ['18.00'],
 ['17.50'],
 ['16.50'],
 ['16.40'],
 ['16.00'],
 ['16.00'],
 ['15.75'],
 ['15.75'],
 ['14.75'],
 ['14.25'],
 ['13.50'],
 ['13.50'],
 ['12.86'],
 ['12.50'],
 ['12.50'],
 ['12.40'],
 ['12.25'],
 ['11.20'],
 ['11.00'],
 ['10.25'],
 ['10.00'],
 ['10.00'],
 ['10.00'],
 ['10.00'],
 ['9.50'],
 ['9.47'],
 ['9.00'],
 ['9.00'],
 ['9.00'],
 ['9.00'],
 ['8.75'],
 ['8.62'],
 ['8.25'],
 ['8.00'],
 ['7.75'],
 ['7.75'],
 ['7.04'],
 ['7.00'],
 ['7.00'],
 ['7.00'],
 ['6.75'],
 ['6.75'],
 ['6.75'],
 ['6.61'],
 ['6.50'],
 ['6.50'],
 ['6.50'],
 ['6.50'],
 ['6.50'],
 ['6.25'],
 ['6.25'],
 ['6.00'],
 ['6.00'],
 ['6.00'],
 ['5.75'],
 ['5.75'],
 ['5.50'],
 ['5.50'],
 ['5.50'],
 ['5.48'],
 ['5.00'],
 ['5.00'],
 ['5.00'],
 ['5.00'],
 ['5.00'],
 ['5.00'],
 ['4.75'],
 ['4.75'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.50'],
 ['4.35'],
 ['4.31'],
 ['4.25'],
 ['4.00'],
 [

In [120]:
date_last = [re.findall(r"[A-z]+\/[0-9]+",row_contents[i].text) for i in range(len(row_contents)) if i%6 == 2]

In [121]:
date_last

[['May/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Mar/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Feb/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['May/19'],
 ['Mar/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['May/19'],
 ['Dec/17'],
 ['Apr/19'],
 ['May/19'],
 ['May/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Jan/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Feb/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['Mar/19'],
 ['May/19'],
 ['May/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['Mar/19'],
 ['Apr/19'],
 ['Dec/18'],
 ['Apr/19'],
 ['Apr/19'],
 ['Mar/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['May/19'],
 ['Apr/19'],
 ['Apr/19'],
 ['Apr/19'],

In [122]:
previous = [re.findall(r"[0-9]+\.[0-9]+",row_contents[i].text) for i in range(len(row_contents)) if i%6 == 3]

In [123]:
previous

[['71.44'],
 [],
 [],
 [],
 ['32.28'],
 [],
 [],
 [],
 ['16.5'],
 ['15.3'],
 [],
 [],
 ['15.75'],
 ['15.75'],
 ['14.75'],
 ['14.25'],
 ['14.5'],
 ['13.5'],
 ['12.86'],
 ['12.5'],
 ['12.5'],
 ['12.4'],
 ['10.75'],
 ['11.5'],
 [],
 ['9.75'],
 [],
 [],
 [],
 [],
 ['9.5'],
 ['9.48'],
 [],
 [],
 [],
 [],
 [],
 ['8.75'],
 ['8.25'],
 [],
 ['7.75'],
 ['7.75'],
 ['7.04'],
 [],
 [],
 [],
 ['6.75'],
 ['6.75'],
 ['6.75'],
 ['6.69'],
 ['6.5'],
 ['6.5'],
 ['6.5'],
 ['6.5'],
 ['6.5'],
 ['6.25'],
 ['6.25'],
 [],
 ['6.25'],
 [],
 ['5.75'],
 ['5.75'],
 ['5.5'],
 ['5.5'],
 ['5.5'],
 ['5.48'],
 [],
 [],
 [],
 [],
 ['5.5'],
 [],
 ['4.75'],
 ['4.75'],
 ['4.5'],
 ['4.5'],
 ['4.5'],
 ['4.5'],
 ['4.5'],
 ['4.5'],
 ['4.5'],
 ['4.75'],
 ['4.5'],
 ['4.5'],
 ['4.35'],
 ['4.34'],
 ['4.25'],
 [],
 ['4.5'],
 [],
 [],
 ['3.75'],
 ['3.55'],
 ['3.5'],
 ['3.5'],
 ['3.5'],
 ['3.5'],
 ['3.5'],
 ['3.5'],
 ['3.5'],
 ['3.53'],
 [],
 [],
 [],
 ['3.25'],
 [],
 [],
 ['2.99'],
 ['3.46'],
 ['2.75'],
 ['2.75'],
 ['2.75'],
 ['2.75']

Okay, this is good enough for now...

We want to go ahead and write this to a csv file...

We can either do this with a loop or we can create a pandas dataframe and use a write csv function.

In [140]:
def write_outp(country, last, date_last, previous):
    with open('scrape_output.csv', 'a') as file:
        for country, last, date_last, previous in zip(country, last, date_last, previous):
            file.write('{},{},{},{}\n'.format(str(country), str(last), str(date_last), str(previous)))
            
write_outp(country, last, date_last, previous)

The idea works well enough!